File: app/src/main/java/com/devdream/cookall/main/explore/ExploreRecipesFragment.java
package com.devdream.cookall.main.explore;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.devdream.cookall.R;
import com.devdream.cookall.core.dto.RecipeDTO;

import java.util.List;

public class ExploreRecipesFragment extends Fragment implements ExploreRecipesListener {

    public static final String TAG = "com.devdream.cookall.main.explore.ExploreRecipesFragment";

    private OnListFragmentInteractionListener mListener;
    private ExploreRecipesPresenter exploreRecipesPresenter;

    private RecyclerView recyclerView;

    public ExploreRecipesFragment() {
        exploreRecipesPresenter = new ExploreRecipesPresenter(this);
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment ExploreRecipesFragment.
     */
    public static ExploreRecipesFragment newInstance() {
        ExploreRecipesFragment fragment = new ExploreRecipesFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Receive here the arguments
//        if (getArguments() != null) {
//        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_explore_recipes_list, container, false);

        // Set the adapter
        if (view instanceof RecyclerView) {
            Context context = view.getContext();
            recyclerView = (RecyclerView) view;

            recyclerView.setLayoutManager(new LinearLayoutManager(context));

            exploreRecipesPresenter.getAllRecipes();
        }
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListFragmentInteractionListener) {
            mListener = (OnListFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void recipesLoaded(List<RecipeDTO> recipes) {
        recyclerView.setAdapter(new ExploreRecipesRecyclerViewAdapter(recipes, mListener));
    }

    public interface OnListFragmentInteractionListener {

        void onListFragmentInteraction(RecipeDTO item);

    }

}

File: app/src/main/java/com/devdream/cookall/main/explore/ExploreRecipesFragment.kt
package com.devdream.cookall.main.explore

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.devdream.cookall.R
import com.devdream.cookall.core.dto.RecipeDTO

class ExploreRecipesFragment : Fragment(), ExploreRecipesListener {

    private var mListener: OnListFragmentInteractionListener? = null
    private val exploreRecipesPresenter: ExploreRecipesPresenter

    private var recyclerView: RecyclerView? = null

    init {
        exploreRecipesPresenter = ExploreRecipesPresenter(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Receive here the arguments
        //        if (getArguments() != null) {
        //        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_explore_recipes_list, container, false)

        // Set the adapter
        if (view is RecyclerView) {
            val context = view.getContext()
            recyclerView = view

            recyclerView!!.layoutManager = LinearLayoutManager(context)

            exploreRecipesPresenter.getAllRecipes()
        }
        return view
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnListFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException("$context must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun recipesLoaded(recipes: List<RecipeDTO>) {
        recyclerView!!.adapter = ExploreRecipesRecyclerViewAdapter(recipes, mListener)
    }

    interface OnListFragmentInteractionListener {

        fun onListFragmentInteraction(item: RecipeDTO)

    }

    companion object {

        val TAG = "com.devdream.cookall.main.explore.ExploreRecipesFragment"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment ExploreRecipesFragment.
         */
        fun newInstance(): ExploreRecipesFragment {
            val fragment = ExploreRecipesFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }

}
